enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+
+### Risks and Mitigations
+
+Currently, all cloud-based CSI drivers shipped by OCP are installed when a cluster is installed on the particular cloud.
+AWS EFS will not be installed by default. We will include a note about opt-in installation in docs.
+
+After installation of the new operators, either as a fresh install or upgrading the existing one, users must create the
+CR when the operator is installed on a new cluster. This is something that was not necessary in the previous releases of
+the operator. We will add a release note and introduce an (info) alert to remind admins to create the CR.
+
+## Design Details
+
+### Rewrite to library-go & ClusterCSIDriver CRD
+
+This should be pretty straightforward. We already ship number of CSI drivers that use library-go. Example usage:
+[AWS EBS CSI driver](https://github.com/openshift/aws-ebs-csi-driver-operator/blob/383a7638b37a4b9a9831c7747e8c499eedcf030f/pkg/operator/starter.go#L67).
+EFS CSI driver operator should not be much more complicated.
+
+Library-go does not support un-installation of CSI driver (and static resources created by `StaticResourcesController`),
+this must be implemented there.
+
+### `SharedVolume` CRD
+
+Since `SharedVolume` is trivial to handle, just creates PV + PVC, we leave handling of the CRD as a implementation
+detail. We may try to copy code from the community operator, but if it's hard to combine library-go and
+controller-runtime in a single binary, we may as well reimplement the handling to library-go style controller.
+
+### Upgrade from the community operator
+
+Prerequisites:
+* The community aws-efs-operator is installed in OCP 4.7 cluster in `openshift-operators` namespace. The CSI driver runs
+in the same namespace too.
+  * The operator `Subscription` contains the community catalog as the operator source.
+
+Expected workflow (confirmed with OLM team):
+
+1. User upgrades to 4.8. Nothing interesting happens at this point, because the operator `Subscription` still points
+   to the community source. The "old" operator is still running.
+2. User edits the `Subscription` and updates the `source:` to the Red Hat certified catalog.
+   (Alternatively, user may un-install the community operator and install the supported one.)
+3. OLM reconciles this change, checks the new catalog and if there is a newer operator there, it will install it as an
+   update. In other words, we need to ship a newer version in the Red Hat catalog than in the community one.
+4. The "new" CSI operator starts. It does not see `ClusterCSIDriver` CR for EFS and only emits Info level alert that
+   there is no CR and does nothing. This means the old CSI driver is still running (installed by the "old" operator)
+   and any apps that use it are still running.
+5. User reads release notes, documentation or alert and creates th CR.
+6. The "new" operator "adopts" the CSI driver objects. Namely, it updates DaemonSet of the CSI driver, which then does
```

Comment:
Doesn't this entail moving several of these objects (the namespace-scoped ones) to a different namespace? If so, not sure there's really a requirement to preserve the same names.

Reply:
I considered moving the driver to `openshift-cluster-csi-drivers` namespace, where AWS EBS CSI driver runs, especially if the operator was installed during cluster installation (having a single predictable namespace helps a lot). We moved to OLM, so using the OLM-native way of managing namespaces is better here. And helps with the migration.

@bertinatto, what do you think? Are we painting ourselves into a corner again?

Added a note that we won't change the namespace.

Reply:
A question about collecting logs with must-gather, is there any difference for putting efs csi driver and operator in different namespaces. If we keep the operator and driver in the old namespace, do we need to create a special image to collect logs for efs csi driver and operator?